onnx/defs/data_type_utils.cc

```diff
@@ -119,25 +119,48 @@ std::string DataTypeUtils::ToString(
     const std::string& right) {
   switch (type_proto.value_case()) {
     case TypeProto::ValueCase::kTensorType: {
-        // Note: We do not distinguish tensors with zero rank (a shape consisting of
-        // an empty sequence of dimensions) here.
-        return left + "tensor(" +
-            ToDataTypeString(type_proto.tensor_type().elem_type()) + ")" +
-            right;
+      // Note: We do not distinguish tensors with zero rank (a shape consisting
+      // of an empty sequence of dimensions) here.
+      return left + "tensor(" +
+          ToDataTypeString(type_proto.tensor_type().elem_type()) + ")" + right;
     }
 #ifdef ONNX_ML
     case TypeProto::ValueCase::kSequenceType: {
-        return ToString(type_proto.sequence_type().elem_type(), left + "seq(", ")" + right);
+      return ToString(
+          type_proto.sequence_type().elem_type(), left + "seq(", ")" + right);
     }
 
     case TypeProto::ValueCase::kMapType: {
-        std::string map_str = "map(" + ToDataTypeString(type_proto.map_type().key_type()) + ",";
-        return ToString(type_proto.map_type().value_type(), left + map_str, ")" + right);
+      std::string map_str =
+          "map(" + ToDataTypeString(type_proto.map_type().key_type()) + ",";
+      return ToString(
+          type_proto.map_type().value_type(), left + map_str, ")" + right);
+    }
+    case TypeProto::ValueCase::kOpaqueType: {
+      static const std::string empty;
+      std::string result;
+      std::string param_str;
+      const auto& op_type = type_proto.opaque_type();
+      const auto param_size = op_type.parameters_size();
+      for (int p = 0; p < param_size; ++p) {
+        const auto& param_proto = op_type.parameters(p);
+        param_str.append(
+            ToString(param_proto, (p == 0) ? empty : ",", empty));
+      }
+      result.append("opq(");
```

Comment:
Just some minor suggestions: why not "opaque" instead of "opq"? Also, not sure if we really need the designations "d()" and "n()" around the domain and name … we can treat them positionally … except I guess that they can be empty …

Reply:
Yes "opaque" is better.

Reply:
I do not mind. Would `sequence` instead of `seq` be a better choice then?

Reply:
keep "sep".